miou/client/dns_client_miou_unix.ml

```diff
@@ -0,0 +1,186 @@
+            Ok (addr, `Tls fd)
+        with End_of_file ->
+            Miou_unix.close fd;
+            error_msgf "Connection to nameservers (via TLS) impossible" end
+    | `Udp ->
+      let is_plaintext = function `Plaintext v -> Either.Left v | _ -> Either.Right () in
+      let[@warning "-8"] (ipaddr, port) :: _, _ = List.partition_map is_plaintext t.nameservers in
+      let proto_number, socket_type = Unix.((getprotobyname "udp").p_proto, SOCK_DGRAM) in
+      let domain = match ipaddr with
+        | Ipaddr.V4 _ -> Unix.PF_INET
+        | Ipaddr.V6 _ -> Unix.PF_INET6 in
+      let fd = Unix.socket domain socket_type proto_number in
+      let addr = Unix.ADDR_INET (Ipaddr_unix.to_inet_addr ipaddr, port) in
+      let connect () =
+        Unix.connect fd addr;
+        ((ipaddr, port), `Udp (Miou_unix.of_file_descr fd)) in
+      match with_timeout ~timeout:t.timeout connect with
+      | Ok value -> Ok value
+      | Error Timeout ->
+          Unix.close fd;
+          error_msgf "Connection to nameservers (via UDP) timeout"
+      | Error exn ->
+          Unix.close fd;
+          error_msgf "Unexpected error: %S" (Printexc.to_string exn)
+  
+  let nameservers { nameservers; proto; _ } = (proto, nameservers)
+  let bind x f = f x
+  let lift = Fun.id
+  let rng = Mirage_crypto_rng.generate ?g:None
+  
+  let connect t =
+    let ( let* ) = Result.bind in
+    let* ((addr, port), fd) = connect_to_nameservers t in
+    Log.debug (fun m -> m "Connected to a nameserver %a:%d" Ipaddr.pp addr port);
+    match fd with
+    | `Tcp _ | `Tls _ -> Ok (`Tcp, { fd; timeout= t.timeout })
+    | `Udp _ -> Ok (`Udp, { fd; timeout= t.timeout })
+  
+  let send_recv_tls ~timeout ~id fd str =
+    let send () = Tls_miou_unix.write fd str in
+    let recv () =
+      let rec go buf rx_len =
+        let expected_len =
+          if rx_len >= 2 then Some (Bytes.get_uint16_be buf 0) else None in
+        match expected_len with
+        | None ->
+          let len = Tls_miou_unix.read fd buf ~off:rx_len in
+          if rx_len + len >= 2 && len > 0 then go buf (rx_len + len)
+          else failwith "TLS connection closed by nameserver"
+        | Some expected_len when rx_len >= expected_len + 2 ->
+          let id' = Bytes.get_uint16_be buf 2 in
+          if id = id'
+          then Bytes.sub_string buf 0 (expected_len + 2)
+          else
+            let buf' = Bytes.make 2048 '\000' in
+            let rx_len' = rx_len - (expected_len + 2) in
+            Bytes.blit buf (expected_len + 2) buf' 0 rx_len';
+            go buf' rx_len'
+        | Some expected_len when Bytes.length buf >= expected_len + 2 ->
+          let len = (expected_len + 2) - rx_len in
+          Tls_miou_unix.really_read fd buf ~off:rx_len ~len;
+          go buf (rx_len + len)
+        | Some expected_len ->
+          let buf' = Bytes.make (expected_len + 2) '\000' in
+          Bytes.blit buf 0 buf' 0 rx_len;
```

Comment:
I'm actually unsure about `buf'` -- what is it's purpose? we allocate and blit into it, but then we throw it away!?

Reply:
I suspect the line below should use `buf'`. The `go` function is a bit hard to read and understand imho...

Reply:
If I read correctly, this is the worst case where the current `buf` is not large enough to contains the actual DNS packet. In this branch, the idea is to allocate a new bytes `buf'`) which can fit the DNS packet and use it (and forget the old `buf`). Actually, we have an error on the `go buf rx_len`, it should be `go buf' rx_len`.

Reply:
Fixed, I added a comment to explain what is the purpose of this branch.